### ISSUE TITLE
[Clang] Confusing error message for const cast between types with the same name

### ISSUE THREAD

**Issue by hnrklssn** [llvm/llvm-project#139677]
```
namespace foo {
  class Type {};
}
const foo::Type *bar();
class Type {};

const_cast<Type*>(bar()); // error: const_cast from 'const Type *' to 'Type *' is not allowed
```

When types have the same name we should emit diagnostics that refer to the fully qualified type names, to disambiguate. I'm sure other diagnostics suffer from the same problem, perhaps the DiagnosticsEngine or something could handle this automatically?

**Comment by shafik** [comment]
This looks ok-ish and consistent w/ other implementations on trunk: https://godbolt.org/z/bMEb3Kh1v

```console
<source>:8:3: error: const_cast from 'const foo::Type *' to 'Type *' is not allowed
    8 |   const_cast<Type*>(bar());
```

**Comment by hnrklssn** [comment]
Hmm, I wonder why I didn't get that 🤔

**Comment by shafik** [comment]
> Hmm, I wonder why I didn't get that 🤔

I am not sure, I was trying older version and don't see it, so if you can reproduce that would help.

**Comment by hnrklssn** [comment]
I was able to reproduce it now, by making the function with the return type also be part of the namespace: https://godbolt.org/z/9YodYeGv9 That way it doesn't explicitly name `foo::` in front of `Type` in its signature. Reproduces on trunk also.

I get a similar thing without the const_cast also, with mismatching constness: https://godbolt.org/z/WaaPrzdrP
However if I match the constness, it does add `(aka 'const foo::Type*')` to the RHS: https://godbolt.org/z/ezjGPMase
